lib/FrontendTool/ReferenceDependencies.h

```diff
@@ -32,9 +32,9 @@ std::vector<std::string>
 reversePathSortedFilenames(const llvm::ArrayRef<std::string> paths);
 
 /// Emit a Swift-style dependencies file for \p SF.
-bool emitReferenceDependenciesIfNeeded(DiagnosticEngine &diags, SourceFile *SF,
-                                       DependencyTracker &depTracker,
-                                       StringRef outputPath);
+bool emitReferenceDependencies(DiagnosticEngine &diags, SourceFile *SF,
+                               DependencyTracker &depTracker,
+                               const std::string &outputPath);
```

Comment:
Yup. What if the string will be used as a null-terminated string? (I assume that's the case here, since it will be passed to a system call eventually.) If the parameter is a std::string (ref), then a caller with a StringRef would have to create a (temporary) std::string to pass in and would pay the execution cost (negligible here) to do the conversion. But a caller with a std::string would just pass it in for free.

OTOH, if this function takes a StringRef, it always has to eventually do the conversion, since the path is passed down to Unix.

But there's likely a piece of the puzzle I'm missing. I'm assuming that a StringRef is not automatically converted to a std::string, so that the execution cost is explicit. I'm also assuming that a StringRef does not remember whether it was created from a null-terminated string or not.
Help me understand what I'm missing. Thanks!

Reply:
It's that last bit: a StringRef does not know whether it came from a null-terminated string. (It would be totally reasonable for it to do so, simply by stealing a bit out of the "size" field, but it doesn't.) That means that if a string ever passes through a StringRef layer in the API, you've lost that information.

In this case, you're right that the string will end up null-terminated because it ends up at a system call. But the API we use to perform that system call is already an LLVM wrapper that takes an arbitrary StringRef, so there's already no way to take advantage of the information. Therefore it's not worth limiting *our* API.

(This is all pretty subtle, so if you want a simpler rule, it's "don't bother optimizing for null-terminated strings at the expense of losing StringRef", because while that *is* a copy and I've been saying we try to avoid copies, it's a copy that can often be done on the stack or on a bump-pointer allocator at the point where it's needed, rather than needing a heap allocation.)